src/rewrites.rs

```diff
@@ -588,3 +712,274 @@ fn check_pat(
         }
     };
 }
+
+/// Struct for storing information on how each pattern maps to its canonical version
+#[derive(Debug)]
+struct MapToCanonical {
+    /// Index into MultiPatterns.canonical_src_pat. Points to the canonical version.
+    index: usize,
+    /// Mapping from variable in this pattern to variable in the canonical pattern.
+    var_map: HashMap<egg::Var, egg::Var>,
+}
+
+/// Struct for the multi-pattern rules. In charge of searching for matches and
+/// applying the rewrite.
+#[derive(Debug)]
+pub struct MultiPatterns {
+    /// Vec of (src_1, src_2, dst_1, dst_2)
+    rules: Vec<(Pattern<Mdl>, Pattern<Mdl>, Pattern<Mdl>, Pattern<Mdl>)>,
+    /// Vec of all unique canonical source patterns (for src_1's and src_2's)
+    canonical_src_pat: Vec<Pattern<Mdl>>,
+    /// Mapping information for each src pattern. The order is the same as in rules
+    src_pat_maps: Vec<(MapToCanonical, MapToCanonical)>,
+}
+
+impl MultiPatterns {
+    /// Construct a MultiPatterns with rules. Each multi-pattern rule contains two matched outputs.
+    ///
+    /// # Parameters
+    ///
+    /// - `rules`: every adjacent pair of entries should belong to the same multi-pattern rule.
+    pub fn with_rules(rules: Vec<&str>) -> MultiPatterns {
+        assert!(rules.len() % 2 == 0);
+
+        let mut multi_rules =
+            Vec::<(Pattern<Mdl>, Pattern<Mdl>, Pattern<Mdl>, Pattern<Mdl>)>::new();
+        let mut canonical_pats = Vec::<Pattern<Mdl>>::new();
+        let mut src_pat_maps = Vec::<(MapToCanonical, MapToCanonical)>::new();
+
+        let mut canonicalize_and_add = |pat: &Pattern<Mdl>| {
+            let (pat_canonical, pat_var_map) = canonicalize(pat);
+            let index_found = canonical_pats.iter().enumerate().find_map(|(i, pattern)| {
+                if *pattern == pat_canonical {
+                    Some(i)
+                } else {
+                    None
+                }
+            });
+            let pat_index = match index_found {
+                Some(i) => i,
+                None => {
+                    canonical_pats.push(pat_canonical);
+                    canonical_pats.len() - 1
+                }
+            };
+            let pat_map = MapToCanonical {
+                index: pat_index,
+                var_map: pat_var_map,
+            };
+            pat_map
```

Comment:
No need to bind `pat_map` you can directly return the struct without assigning it to a variable.

Reply:
Done